pkg/authentication/oidc/authenticate.go

```diff
@@ -0,0 +1,126 @@
+// an OIDC/OAuth2 Authorization Code Flow.
+// Use [WithCodeFlow] for implementation.
+type CodeFlowAuthentication[T Ctx[C, S], C oidc.IDClaims, S rp.SubjectGetter] struct {
+	relyingParty rp.RelyingParty
+}
+
+// WithCodeFlow creates the OIDC/OAuth2 Authorization Code Flow implementation of the [authentication.Handler] interface.
+// The token endpoint itself requires some [ClientAuthentication] of the client.
+// Possible implementation are [PKCEAuthentication] and [ClientIDSecretAuthentication].
+func WithCodeFlow[T Ctx[C, S], C oidc.IDClaims, S rp.SubjectGetter](auth ClientAuthentication) authentication.HandlerInitializer[T] {
+	return func(ctx context.Context, domain string) (authentication.Handler[T], error) {
+		relyingParty, err := auth(ctx, domain)
+		if err != nil {
+			return nil, err
+		}
+		return &CodeFlowAuthentication[T, C, S]{
+			relyingParty: relyingParty,
+		}, nil
+	}
+}
+
+type ClientAuthentication func(ctx context.Context, domain string) (rp.RelyingParty, error)
+
+// PKCEAuthentication allows to authenticate the code exchange request with Proof Key of Code Exchange (PKCE).
+func PKCEAuthentication(clientID, redirectURI string, scopes []string, cookieHandler *httphelper.CookieHandler) ClientAuthentication {
+	return func(ctx context.Context, domain string) (rp.RelyingParty, error) {
+		return newRP(ctx, domain, clientID, "", redirectURI, scopes, rp.WithPKCE(cookieHandler))
+	}
+}
+
+// ClientIDSecretAuthentication allows to authenticate the code exchange request with client_id and client_secret provide by ZITADEL.
+func ClientIDSecretAuthentication(clientID, clientSecret, redirectURI string, scopes []string, cookieHandler *httphelper.CookieHandler) ClientAuthentication {
+	return func(ctx context.Context, domain string) (rp.RelyingParty, error) {
+		return newRP(ctx, domain, clientID, clientSecret, redirectURI, scopes, rp.WithCookieHandler(cookieHandler))
+	}
+}
+
+// DefaultAuthentication is a short version of [WithCodeFlow[*UserInfoContext[*oidc.IDTokenClaims, *oidc.UserInfo], *oidc.IDTokenClaims, *oidc.UserInfo]]
+// with the client_id, redirectURI and encryptionKey and optional scopes.
+// If no scopes are provided, `"openid", "profile", "email"` will be used.
+func DefaultAuthentication(clientID, redirectURI string, key string, scopes ...string) authentication.HandlerInitializer[*UserInfoContext[*oidc.IDTokenClaims, *oidc.UserInfo]] {
+	if len(scopes) == 0 {
+		scopes = []string{oidc.ScopeOpenID, oidc.ScopeProfile, oidc.ScopeEmail}
+	}
+	return WithCodeFlow[*UserInfoContext[*oidc.IDTokenClaims, *oidc.UserInfo], *oidc.IDTokenClaims, *oidc.UserInfo](
+		PKCEAuthentication(clientID, redirectURI, scopes, httphelper.NewCookieHandler([]byte(key), []byte(key))),
+	)
+}
+
+func newRP(ctx context.Context, domain, clientID, clientSecret, redirectURI string, scopes []string, options ...rp.Option) (rp.RelyingParty, error) {
+	if len(scopes) == 0 {
+		scopes = []string{oidc.ScopeOpenID}
```

Comment:
List of default scopes "openid email profile" for all functions?

Reply:
as this function will always be wrapped, the intention was to make sure the RP at least works